Documents/AnalyticsToolDesignDocument.md

```diff
@@ -0,0 +1,108 @@
+- The **file system** must be carefully considered in the case of working with a mounted repository. VFS4G intercepts file I/O calls at the kernel level to inject its own behavior (the projected file system and if needed a network call to pull down the file which the system asked for in the correct format). Because of this, enumerating directories can cause additional data to be pulled down from the remote by hydrating extra files or further hydrating those that already exist on disk. Accordingly, care must be taken when interacting with the file system lest the tool designed to help users increase enlistment heath end up causing more problems.
+
+## Components
+
+- **API** - The server component of the client-server relationship. Responsible for the logic and computation of statistical data, as well as the assembly and sending of a **json statistics object** over the named pipe to the client. It sits within the `InProgressMount` process and interacts with the VFS4G internals
+- **Verb** - The `gvfs statistics` verb will be the component that handles the client side interaction with the API. When run, it will collect and store the statistics from the enlistment to either be displayed immediately or collected for trend purposes to then provide understanding in the future. This is also the point where data can be sent to telemetry for our internal analysis. The verb will also output the data to the console with a tool to help visualize and understand the data. It will output the analytics data to the command line and will also be used during the design period to test the API and verb
+- **Utility tools for dehydration** - *Stretch goal* | After the analytics are collected, these components will provide utilities to help the user get their enlistment back to a healthier state. The ideal goal is to have VFS4G suggest uses to the user (EG: `This directory was hydrated by Sublime text editor. You can dehydrate it by running "gvfs dehydrate --confirm -d /path/to/repo"`)
+- **Additional visualization tools** - *Stretch goal* | With a comprehensive API in place, the possibility opens up for different visualization tools to be created that consume the json data and display it to the user in different ways, potentially even eventually interfacing with the utility tools for an interactive GUI
+
+## Example of CLI Tool Usage
+
+```terminal
+$ gvfs statistics
+
+Repository statistics
+Total paths tracked by git:        3,123,456 |  100%
+Total number of placeholders:        123,456 |    4%
+Total number of modififed paths:       1,234 |   <1%
+
+Total hydration percentage:                       4%
+Most hydrated directories:
+  98% | src/main    | Primarily hydrated by "Sublime Text Editor"
+  79% | src/util    | Primarily hydrated by "Visual Studio 2019"
+  58% | assets/img  | Primarily hydrated by "cmd.exe"
+
+Repository status:  Healthy
+Recommended action: None
+
+To view git command performance run with "--perf"

```

Comment:
We will want to think about progress indicators if we are going a `git` perf benchmark. Love the idea!

Reply:
I wonder where this information will be stored because VFSForGit does not have the complete picture since it only knows about commands that take the lock and might not be a total picture of the time.  We should see if trace2 or something @jeffhostetler has done can help give us this information.

Reply:
Oh, I was thinking we'd have some set of benchmark commands that we run as part of the `statistics` verb. The progress indicator would say "Running `git log --graph -n 10`" or something as we run each command. It's important to find a set of read-only commands to test.

Reply:
Having a user self-serve certain Git statistics from their trace2 logs could be interesting, especially for external customers.

Reply:
> Having a user self-serve certain Git statistics from their trace2 logs could be interesting, especially for external customers.

One idea that has been discussed - the Git service daemon has the ability to dispatch data to multiple subscribers. I think it could dump trace2 logs (or some version of them) to disk, as well as whatever else it does. This would not address external scenarios (maybe for external use scenarios git can be configured to write out the trace2 logs, instead of going through a broker service).